### PR TITLE
Add project page

### PR DESCRIPTION
**Pull request by kuranovandry**

### CODE REVIEW

File: app/controllers/projects_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  def new
+    @project = Project.new
+  end
+
+  def create
+    @project = Project.new(project_params)
+
+    if @project.save
+      flash[:success] = 'Project was successfully created.'
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Move notice to locales

File: app/controllers/projects_controller.rb [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      redirect_to projects_path
+    else
+      render 'new'
+    end
+  end
+
+  def edit
+    @project = Project.find(params[:id])
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Extract to before_filter

File: app/controllers/projects_controller.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def update
+    @project = Project.find(params[:id])
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Extract to before_filter

File: app/controllers/projects_controller.rb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    if @project.update(project_params)
+      flash[:success] = 'Project was successfully updated.'
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Move notice to locales

File: app/controllers/projects_controller.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+      redirect_to @project
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Use rails url helpers

File: app/controllers/projects_controller.rb [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    else
+      render 'edit'
+    end
+  end
+
+  def show
+    @project = Project.find(params[:id])
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Extract to before_filter

File: app/controllers/projects_controller.rb [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  end
+
+  def destroy
+    @project = Project.find(params[:id])
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Extract to before_filter

File: app/controllers/projects_controller.rb [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    @project.destroy
+
+    flash[:success] = 'Project was successfully destroyed.'
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Move notice to locales

File: app/models/project.rb [hunk 9]
```diff
@@ -0,0 +1,5 @@
+class Project < ActiveRecord::Base
+
+  belongs_to :user
```
Comment (maksymst) [inline app/models/project.rb:6]: Add validations

File: app/views/projects/_form.html.haml [hunk 10]
```diff
@@ -0,0 +1,25 @@
+= form_for(@project, :html => { :class => 'form_horizontal'}) do |f|
+  .form-field-group
+    = f.label :name, :class => 'col-sm-2 control-label'
```
Comment (maksymst) [inline app/views/projects/_form.html.haml]: use new hash style

File: app/views/projects/show.html.haml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+.col-md-9
+  %strong Name:
+  = @project.name
+.col-md-9
+  %strong Description:
+  = @project.description
+.col-md-9
+  %strong Start Date:
+  = @project.start_date
```
Comment (maksymst) [inline app/views/projects/show.html.haml]: use `strftime` to format date and time

File: app/views/shared/_navigation.html.haml [hunk 12]
```diff
@@ -2,7 +2,11 @@
   .container
     .navbar-header
       %a.navbar-brand{href: "/"}
-        %b Devise project
+        %b Users
+      %a.navbar-brand{href: "/projects"}
```
Comment (maksymst) [inline app/views/shared/_navigation.html.haml]: Use rails url helpers

File: db/migrate/20150428130619_create_projects.rb [hunk 13]
```diff
@@ -0,0 +1,13 @@
+class CreateProjects < ActiveRecord::Migration
+  def change
+    create_table :projects do |t|
+      t.string :name
+      t.string :description
+      t.integer :user_id
```
Comment (maksymst) [inline db/migrate/20150428130619_create_projects.rb]: use `t.belongs_to :user` + index

File: db/migrate/20150428130619_create_projects.rb [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+      t.datetime   :start_date
```
Comment (maksymst) [inline db/migrate/20150428130619_create_projects.rb]: why do you use `datetime` for date?

File: app/views/projects/index.html.haml [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+      %th Start date
+      %th End date
+      %th Date Publised
+      %th{:align => "right", :colspan => "2"} Options
+    - @projects.each do |project|
+      %tr{"data-link" => project_path(project)}
+        %td= project.name
+        %td= project.start_date
+        %td= project.end_date
+        %td= button_to 'Edit', edit_project_path(project), method: :get, :class => 'btn btn-info'
```
Comment (maksymst) [inline app/views/projects/index.html.haml]: use `link_to`

File: app/views/projects/index.html.haml [hunk 16]
```diff
@@ -0,0 +1,17 @@
...
+        %td= button_to 'Destroy', project_path(project), method: :delete, data: { confirm: 'Are you sure?' }, :class => 'btn btn-danger'
```
Comment (maksymst) [inline app/views/projects/index.html.haml]: use link_to

File: app/controllers/projects_controller.rb [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+  def index
+    @projects = Project.all
+  end
+
+  def new
+    @project = Project.new
+  end
+
+  def create
+    # binding.pry
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: remove

File: app/controllers/projects_controller.rb [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    else
+      render 'new'
+    end
+  end
+
+  def update
+    if @project.update(project_params)
+      flash[:success] = t(:success_update)
+      redirect_to projects_path
+
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: extra blank line

File: app/controllers/projects_controller.rb [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+  def destroy
+    @project.destroy
+    flash[:success] = t(:success_destroy)
+    redirect_to projects_path
+  end
+
+  private
+
+  def project_params
+    params.require(:project).permit(:name, :description, :end_date, :start_date, :user_id)
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: why did you include `:user_id`?

File: app/views/projects/index.html.haml [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+      %th Start date
+      %th End date
+      %th Date Publised
+      %th{align: "right", colspan: "2"} Options
+    - @projects.each do |project|
+      %tr{"data-link" => project_path(project)}
+        %td= project.name
+        %td= project.start_date
+        %td= project.end_date
+        %td= link_to 'Edit', edit_project_path(project), method: :get, class: 'btn btn-info'
```
Comment (maksymst) [inline app/views/projects/index.html.haml]: `method: :get` is redundant.

File: db/migrate/20150428130619_create_projects.rb [hunk 21]
```diff
@@ -0,0 +1,14 @@
+class CreateProjects < ActiveRecord::Migration
+  def change
+    create_table :projects do |t|
+      t.string :name
+      t.string :description
+      t.date   :start_date
+      t.date   :end_date
+      t.belongs_to :user
```
Comment (maksymst) [inline db/migrate/20150428130619_create_projects.rb]: `belongs_to` supports :index option

File: db/migrate/20150428130619_create_projects.rb [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+
+      t.timestamps null: false
+    end
+    add_index :projects, :user_id
```
Comment (maksymst) [inline db/migrate/20150428130619_create_projects.rb]: see comment above

File: spec/controllers/projects_controller_spec.rb [hunk 23]
```diff
@@ -0,0 +1,32 @@
+require 'rails_helper'
+
+RSpec.describe UsersController, type: :controller do
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: Where are necessary tests?

File: app/models/project.rb [hunk 24]
```diff
@@ -0,0 +1,6 @@
+class Project < ActiveRecord::Base
+
+  validates_presence_of :name, :description, :start_date, :end_date, :user_id
```
Comment (maksymst) [inline app/models/project.rb]: Organize your model settings with comments and blank lines. It is not requirement but is is very useful. See example

``` ruby
class Channel < ActiveRecord::Base
  RECOMMENDED_SIZE = 5

#-------------Plugins------------------
  acts_as_followable

#-------------Associations-------------
  belongs_to :owner, class_name: 'User'
  belongs_to :position_department, class_name: 'Department'

  has_one :location, as: :owner, dependent: :destroy

  has_many :position_types, through: :position_department
  has_many :updates, as: :postable, class_name: 'Post', dependent: :destroy

#-------------Scopes-------------------
  scope :by_professions, -> (position_type_ids) { joins(:position_types).where(position_types: { id: position_type_ids }) }
  scope :by_location, -> (*location) { where(id: Location.near(*location).where(owner_type: 'Group').map(&:owner_id)) }

  accepts_nested_attributes_for :location

#-------------Callbacks----------------
  after_create :create_header
  after_save :check_moderator

#-------------Class methods------------
  class << self
    def recommended(user)
      channels = initial_recommended(user).limit(RECOMMENDED_SIZE)
      channels += fullfill_recommended(channels) if channels.size < RECOMMENDED_SIZE
      channels
    end

    private

    def fullfill_recommended(channels)
      where.not(id: channels.map(&:id)).order('RANDOM()').limit(RECOMMENDED_SIZE - channels.size)
    end

    def initial_recommended(user)
      profession_ids = user.professions.pluck(:id)

      if profession_ids.any?
        by_professions(profession_ids)
      else
        order('RANDOM()')
      end
    end
  end

#-------------Instance methods---------
  private

  def check_moderator
    owner.follow self if owner && !owner.following?(self)
  end

  def create_header
    create_logo
  end
end
```

File: app/controllers/projects_controller.rb [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+  end
+
+  def new
+    @project = Project.new
+  end
+
+  def create
+    # binding.pry
+    @project = Project.new(project_params)
+    @project.user_id = current_user.id
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: Use association to create such records.

File: app/controllers/projects_controller.rb [hunk 26]
```diff
@@ -0,0 +1,50 @@
...
+    @projects = Project.all
+  end
+
+  def new
+    @project = Project.new
+  end
+
+  def create
+    @project = Project.new(project_params)
+    @project.user_id = current_user.id
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: here was comment about associations

File: app/controllers/projects_controller.rb [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  private
+
+  def project_params
+    params.require(:project).permit(:name, :description, :end_date, :start_date)
+  end
+
+  def get_project
+    @project = Project.find(params[:id])
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: SECURITY ISSUE: If you write like this you can view or edit any project in system.

File: app/views/projects/index.html.haml [hunk 28]
```diff
@@ -0,0 +1,16 @@
...
+      %th Start date
+      %th End date
+      %th{align: "right", colspan: "2"} Options
+    - @projects.each do |project|
+      %tr{"data-link" => project_path(project)}
+        %td= project.name
+        %td= project.start_date.strftime("%e/%m/%Y at %w:%M %p")
+        %td= project.end_date.strftime("%e/%m/%Y at %w:%M %p")
+        %td= link_to 'Destroy', project_path(project), method: :delete, data: { confirm: 'Are you sure?' }, class: 'btn btn-danger'
+        %td= link_to 'Edit', edit_project_path(project), method: :get, class: 'btn btn-info' if project.user_id == current_user.id
```
Comment (maksymst) [inline app/views/projects/index.html.haml]: move condition to helper and remove `method: :get`

File: spec/controllers/projects_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,92 @@
...
+        expect(response).to render_template(:index)
+      end
+    end
+  end
+
+  describe "DELETE #destroy" do
+    context 'user project owner' do
+      it 'removes project' do
+        delete :destroy, id: projects.id
+        binding.pry
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: wtf?

File: spec/controllers/projects_controller_spec.rb [hunk 30]
```diff
@@ -0,0 +1,92 @@
...
+        expect(response).to be_success
+      end
+    end
+    context 'user not project owner' do
+      it 'does not remove project' do
+        user = create(:user, id: 6)
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: wtf?

File: spec/controllers/projects_controller_spec.rb [hunk 31]
```diff
@@ -0,0 +1,92 @@
...
+        expect(response).to be_false
+      end
+    end
+  end
+  
+  describe "PUT #update" do
+    context 'user project owner' do
+      it 'update project' do
+        delete :update, id: projects.id
+        binding.pry
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: wtf?

File: spec/controllers/projects_controller_spec.rb [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+        expect(response).to be_success
+      end
+    end
+    context 'user not project owner' do
+      it 'does not update project' do
+        user = create(:user, id: 6)
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: wtf?

File: spec/controllers/projects_controller_spec.rb [hunk 33]
```diff
@@ -0,0 +1,92 @@
...
+        expect(response).to be_false
+      end
+    end
+  end
+
+  describe "GET #show" do
+    before :each do
+      get :show, id: projects
+    end
+
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb:89]: extra line

File: spec/controllers/projects_controller_spec.rb [hunk 34]
```diff
@@ -0,0 +1,92 @@
...
+      end
+
+      it 'renders index' do
+        expect(response).to render_template(:index)
+      end
+    end
+  end
+
+  describe "DELETE #destroy" do
+    context 'user project owner' do
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: unclear context

File: spec/controllers/projects_controller_spec.rb [hunk 35]
```diff
@@ -0,0 +1,92 @@
...
+      it 'removes project' do
+        delete :destroy, id: projects.id
+        binding.pry
+        expect(response).to be_success
+      end
+    end
+    context 'user not project owner' do
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: missing word in context

File: spec/support/controller_helpers.rb [hunk 36]
```diff
@@ -1,12 +1,11 @@
 module ControllerHelpers
```
Comment (maksymst) [inline spec/support/controller_helpers.rb:1]: changes in this file are not necessary

File: app/views/projects/index.html.haml [hunk 37]
```diff
@@ -0,0 +1,16 @@
...
+      %th Name
+      %th Start date
+      %th End date
+      %th{align: "right", colspan: "2"} Options
+    - @projects.each do |project|
+      %tr{"data-link" => project_path(project)}
+        %td= project.name
+        %td= project.start_date.strftime("%e/%m/%Y at %w:%M %p")
+        %td= project.end_date.strftime("%e/%m/%Y at %w:%M %p")
+        %td= link_to 'Destroy', project_path(project), method: :delete, data: { confirm: 'Are you sure?' }, class: 'btn btn-danger'
```
Comment (maksymst) [inline app/views/projects/index.html.haml]: this link is accessible by every user

File: app/models/project.rb [hunk 38]
```diff
@@ -0,0 +1,9 @@
+class Project < ActiveRecord::Base
+  #-----------Validations------------
+  validates_presence_of :name, :description, :start_date, :end_date
+  
+  #-------------Associations-------------
+  belongs_to :user
+
```
Comment (maksymst) [inline app/models/project.rb]: extra lines

File: spec/controllers/projects_controller_spec.rb (hunk 32 above)
Comment (kuranovandry) [inline spec/controllers/projects_controller_spec.rb]: I create user with wrong id. Project have user_id 5. Here i create user with id 6 to create error

File: spec/controllers/projects_controller_spec.rb (hunk 30 above)
Comment (kuranovandry) [inline spec/controllers/projects_controller_spec.rb]: I create user with wrong id. Project have user_id 5. Here i create user with id 6 to create error

File: app/controllers/projects_controller.rb [hunk 39]
```diff
@@ -0,0 +1,48 @@
...
+  end
+
+  private
+
+  def project_params
+    params.require(:project).permit(:name, :description, :end_date, :start_date)
+  end
+
+  def get_project
+    @project = current_user.projects.find(params[:id])
```
Comment (maksymst) [inline app/controllers/projects_controller.rb:52]: this means you can view only own projects which is incorrect

File: spec/support/controller_helpers.rb (hunk 36 above)
Comment (kuranovandry) [inline spec/support/controller_helpers.rb:1]: With this changes corrent_user method works properly on rspec test

File: config/locales/en.yml [hunk 40]
```diff
@@ -21,3 +21,6 @@
 
 en:
   hello: "Hello world"
+  success_create: 'Project was successfully created.'
```
Comment (maksymst) [inline config/locales/en.yml]: items related to projects are not separated. it prevents adding more locals with same keys.
also locals keys are incorrect
replace `success_create` with `successful_create`. this have more sense

File: spec/controllers/projects_controller_spec.rb (hunk 35 above)
Comment (MaximSt) [inline spec/controllers/projects_controller_spec.rb]: add new line

File: spec/controllers/projects_controller_spec.rb [hunk 41]
```diff
@@ -0,0 +1,92 @@
...
+  end
+
+  describe "DELETE #destroy" do
+    context 'user project owner' do
+      it 'removes project' do
+        delete :destroy, id: projects.id
+        binding.pry
+        expect(response).to be_success
+      end
+    end
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb:51]: add new line after

File: spec/controllers/projects_controller_spec.rb [hunk 42]
```diff
@@ -0,0 +1,89 @@
+require 'rails_helper'
+
+RSpec.describe ProjectsController, type: :controller do
+  let!(:project) { create(:project) }
+  let!(:user) { create(:user) }
+  let!(:projects) { create(:project, user: user)}
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: why variable name is plural?

File: spec/controllers/projects_controller_spec.rb [hunk 43]
```diff
@@ -0,0 +1,89 @@
...
+      it 'update project' do
+        delete :update, id: projects.id
+        expect(response).to be_success
+      end
+    end
+    context 'when user is not project owner' do
+      it 'does not update project' do
+        user = create(:user)
+        login_with(user)
+        binding.pry
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: wtf?

File: spec/controllers/projects_controller_spec.rb [hunk 44]
```diff
@@ -0,0 +1,89 @@
...
+  describe "PUT #update" do
+    context 'when user is project owner' do
+      it 'update project' do
+        delete :update, id: projects.id
+        expect(response).to be_success
+      end
+    end
+    context 'when user is not project owner' do
+      it 'does not update project' do
+        user = create(:user)
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: what the purpose of next two lines?

File: spec/controllers/projects_controller_spec.rb [hunk 45]
```diff
@@ -0,0 +1,95 @@
...
+      end
+      it { expect(response).to redirect_to(projects_path) }
+      it { expect(project[:name]).to eql attr[:name] }
+    end
+
+    context 'when user is not project owner' do
+      before (:each) do
+        put :update, id: another_project.id, project: attr
+        project.reload
+      end
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb:81]: add new line

File: spec/controllers/projects_controller_spec.rb [hunk 46]
```diff
@@ -0,0 +1,95 @@
...
+      it { expect(response).to redirect_to(projects_path) }
+      it { expect(project[:name]).to eql attr[:name] }
+    end
+
+    context 'when user is not project owner' do
+      before (:each) do
+        put :update, id: another_project.id, project: attr
+        project.reload
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: why do you need this?

File: spec/controllers/projects_controller_spec.rb [hunk 47]
```diff
@@ -0,0 +1,95 @@
...
+      end
+      it { expect(response).to redirect_to(projects_path) }
+      it { (expect(flash[:notice])).to be_present }
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: check for expected flash message

File: spec/controllers/projects_controller_spec.rb [hunk 48]
```diff
@@ -0,0 +1,95 @@
...
+      it 'does not remove project' do
+        expect {
+          delete :destroy, id: another_project.id
+        }.to change { Project.count }.by(0)
+      end
+    end
+  end
+
+  describe "PUT #update" do
+    let(:attr) { { name: 'new name' } }
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb:64]: add new line

File: config/locales/en.yml [hunk 49]
```diff
@@ -20,4 +20,8 @@
 # available at http://guides.rubyonrails.org/i18n.html.
 
 en:
-  hello: "Hello world"
+  project:
+    successful_create: 'Project was successfully created.'
+    successful_update: 'Project was successfully updated.'
+    successful_destroy:  'Project was successfully destroyed.'
+    error_destroy: 'Error'
```
Comment (maksymst) [inline config/locales/en.yml]: wtf?

File: app/controllers/projects_controller.rb [hunk 50]
```diff
@@ -0,0 +1,59 @@
...
+  private
+
+  def project_params
+    params.require(:project).permit(:name, :description, :end_date, :start_date)
+  end
+
+  def get_project
+    begin
+      @project = current_user.projects.find(params[:id])
+    rescue
```
Comment (maksymst) [inline app/controllers/projects_controller.rb]: why do you catch exception here?

File: spec/controllers/projects_controller_spec.rb [hunk 51]
```diff
@@ -0,0 +1,98 @@
...
+
+      it { expect(response).to redirect_to(projects_path) }
+      it { expect(project[:name]).to eql attr[:name] }
+    end
+
+    context 'when user is not project owner' do
+      it 'does not update project' do
+        expect {
+          put :update, id: another_project.id
+        }.to raise_exception
```
Comment (maksymst) [inline spec/controllers/projects_controller_spec.rb]: what kind of exception?

**Pull request merged**